### PR TITLE
Quickstart: use SSL for all network communication: OpenSearch and Jupyter

### PR DESCRIPTION
**Pull request by alexaryn**
Using self-generated, self-signed certificates.  No longer using mini HTTP server for signaling; using OpenSearch cluster metadata instead.

### CODE REVIEW

File: opensearch/sycamore-opensearch.sh [hunk 1]
```diff
@@ -427,4 +433,61 @@ setup_transient() {
...
+    echo "Created admin certificate"
+
+    rm -f node-req.pem admin-req.pem
+
+    ${OLDMASK}
+    popd
+}
+
+setup_security() {
+    # Set up security plugin configuration
```
Comment (eric-anderson) [inline opensearch/sycamore-opensearch.sh:483]: Can you add some comments on what this is doing, e.g. -icl -nhnv 
I also don't understand the -cd; I would have guessed it cds into that directory, but then the cert paths would all be wrong since they wouldn't be under that directory.

File: opensearch/sycamore-opensearch.sh [hunk 2]
```diff
@@ -427,4 +433,61 @@ setup_transient() {
...
+    plugins/opensearch-security/tools/securityadmin.sh \
+    -cd config/opensearch-security -icl -nhnv -cacert config/cacert.pem \
+    -cert config/admin-cert.pem -key config/admin-key.pem
+
+    # Semaphore to signal completion
+    _curl -X PUT "${BASE_URL}/_cluster/settings" -o /dev/null --json \
+    '{"persistent":{"cluster":{"metadata":{"aryn_ssl_setup_complete":1}}}}'
```
Comment (eric-anderson) [inline opensearch/sycamore-opensearch.sh:490]: Does ssl support survive restarts or do you need to run the securityadmin command each time?
Where are you using the ssl_setup_complete semaphore?  I only see it here and not above.

For the semaphores should we have them be versioned, so use something like :20240206?  Then if we have to update the config we can tell what setup it was at (we could also put that off until the first time we have to do it).

File: opensearch/roles_mapping.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+_meta:
+  type: rolesmapping
+  config_version: 2
+
+all_access:
+  backend_roles:
+  - opendistro_security_anonymous_backendrole
```
Comment (eric-anderson) [inline opensearch/roles_mapping.yml:7]: This is magically created by the security setup?
We're going to get a username/password in a followup PR?

File: opensearch/Dockerfile [hunk 4]
```diff
@@ -24,4 +24,12 @@ LABEL git_branch=${GIT_BRANCH}
 LABEL git_commit=${GIT_COMMIT}
 LABEL git_diff=${GIT_DIFF}
 
-CMD ./sycamore-opensearch.sh
+USER root
+RUN yum install -yq openssl jq && yum clean all
+RUN chown 1000:1000 \
```
Comment (eric-anderson) [inline opensearch/Dockerfile:29]: Why are you having to chown here?  They were installed above as 1000.

File: opensearch/Dockerfile [hunk 5]
```diff
@@ -24,4 +24,12 @@ LABEL git_branch=${GIT_BRANCH}
...
+    /usr/share/opensearch/sycamore-opensearch.sh \
+    /usr/share/opensearch/config/*.yml
+RUN chmod 600 /usr/share/opensearch/config/*.yml
```
Comment (eric-anderson) [inline opensearch/Dockerfile:32]: Does the chmod matter? I have learned that Docker is a bit stupid about stuff like this and near as I can tell this might just result in a layer which is a copy of the thing (that's what happens with chown).  Doesn't really matter given how large our images are, but seems better to not do if possible.

**Review by eric-anderson — CHANGES_REQUESTED** [review]

File: importer/docker_local_import.py [hunk 6]
```diff
@@ -244,11 +244,10 @@ def import_files(root, files):
 
 def wait_for_opensearch_ready():
     print("Waiting for opensearch to become ready...", end="")
-    # magic port number needs to match the status port from the aryn-opensearch.sh docker script
     while True:
         try:
-            r = requests.get("http://opensearch:43477/statusz")
-            if r.status_code == 200 and r.text == "OK\n":
+            r = requests.get("https://opensearch:9200/_cluster/settings", verify=False)
+            if r.status_code == 200 and "aryn_deploy_complete" in r.text:
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply importer/docker_local_import.py:250]: I think this does persist.  OpenSearch seems to offer a transient option that resets across restarts.  Good catch.

File: opensearch/sycamore-opensearch.sh [hunk 7]
```diff
@@ -38,14 +37,9 @@ main() {
...
-    echo "OK" >"${http_status_file}"
-    python3 -m http.server -d http_serve "${http_status_port}" -d "${http_server_dir}" \
-            >/tmp/http_server.log 2>&1 &
-    local pid=$!
-    echo "${pid}" >/tmp/http_server.pid
-    trap "kill -TERM $(cat /tmp/http_server.pid)" EXIT
-    disown "${pid}" # don't wait for it to exit
+    # Semaphore to signal completion
+    _curl -X PUT "${BASE_URL}/_cluster/settings" -o /dev/null --json \
+    '{"persistent":{"cluster":{"metadata":{"aryn_deploy_complete":1}}}}'
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:42]: I think we can just switch this to a transient setting and then the check wouldn't be needed, right?

[the remark this answers is not in the document]
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:42]: By the way, why is the stuff in setup_transient not persistent?  Henry seems to think registered models survive OpenSearch restarts.

File: crawler/http/Dockerfile [hunk 8]
```diff
@@ -1,6 +1,6 @@
 # Repo name: arynai/sycamore-crawler-http
 
-FROM python:3
+FROM python:3.11
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply crawler/http/Dockerfile:3]: Good point.  Since PyTorch isn't compatible with 3.11 and the UI specifies 3.11, this seemed the saner route, and it requires one less Docker image on my laptop.

File: importer/docker_local_import.py [hunk 9]
```diff
@@ -374,7 +373,7 @@ def get_os_client_args():
         "hosts": [{"host": "localhost", "port": 9200}],
         "http_compress": True,
         "http_auth": ("admin", "admin"),
-        "use_ssl": False,
+        "use_ssl": True,
         "verify_certs": False,
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply importer/docker_local_import.py:377]: It's documented that we're doing self-signed certificates at this stage.

File: jupyter/run-jupyter.sh [hunk 10]
```diff
@@ -16,6 +16,13 @@ ln -snf "${JUPYTER_CONFIG_DOCKER}" $HOME/.jupyter
 
 rm /app/.local/share/jupyter/runtime/jpserver-*-open.html 2>/dev/null
 
+: ${HOST:=localhost}
+openssl req -batch -x509 -newkey rsa:4096 -days 10000 \
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply jupyter/run-jupyter.sh:20]: Seemed like RSA was more widely deployed.  I didn't test with the other type.

File: opensearch/Dockerfile [hunk 11]
```diff
@@ -24,4 +24,12 @@ LABEL git_branch=${GIT_BRANCH}
 LABEL git_commit=${GIT_COMMIT}
 LABEL git_diff=${GIT_DIFF}
 
-CMD ./sycamore-opensearch.sh
+USER root
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply opensearch/Dockerfile:27]: I wanted to limit to one root block.  The chown/chmod stuff needs to be after the file copying.  Do you think two root blocks are a better idea?

File: jupyter/run-jupyter.sh [hunk 12]
```diff
@@ -16,6 +16,13 @@ ln -snf "${JUPYTER_CONFIG_DOCKER}" $HOME/.jupyter
...
+-subj "/C=US/ST=California/O=Aryn.ai/CN=${HOST}" \
+-extensions v3_req -addext "subjectAltName=DNS:${HOST}" \
+-noenc -keyout "/app/${HOST}-key.pem" -out "/app/${HOST}-cert.pem" 2> /dev/null
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply jupyter/run-jupyter.sh:23]: I'll move them.

File: opensearch/sycamore-opensearch.sh [hunk 13]
```diff
@@ -427,4 +433,61 @@ setup_transient() {
...
+
+    pushd config
+    local OLDMASK=$(umask -p)
+    umask 077
+
+    # 1. Make fake certificate authority (CA) certificate.  OpenSearch
+    # requires a root certificate to be specified.
+    openssl req -batch -x509 -newkey rsa:4096 -days "${DAYS}" \
+    -subj "/C=US/ST=California/O=Aryn.ai/CN=Fake CA" \
+    -extensions v3_ca -noenc -keyout cakey.pem -out cacert.pem 2> /dev/null
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:448]: For now, it doesn't matter because nothing is keeping a list of accepted OpenSearch certificates, unlike browsers and HTTPS.  Indeed, the security plugin needs them to be in config.  I can give the symlinks a try; a quick web search suggests it will work.

File: opensearch/sycamore-opensearch.sh [hunk 14]
```diff
@@ -427,4 +433,61 @@ setup_transient() {
...
+    local HOST="${SSL_HOSTNAME:-localhost}"
+    local DAYS=10000
+
+    pushd config
+    local OLDMASK=$(umask -p)
+    umask 077
+
+    # 1. Make fake certificate authority (CA) certificate.  OpenSearch
+    # requires a root certificate to be specified.
+    openssl req -batch -x509 -newkey rsa:4096 -days "${DAYS}" \
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:446]: I would not expect them to fail unless the disk is full.  I can add a failsafe, though.

File: opensearch/sycamore-opensearch.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:448]: Good idea.  Done.

File: opensearch/sycamore-opensearch.sh [hunk 15]
```diff
@@ -427,4 +433,61 @@ setup_transient() {
...
+    -extensions v3_req -addext "basicConstraints=critical,CA:FALSE" \
+    -noenc -keyout admin-key.pem -out admin-req.pem 2> /dev/null
+
+    # 3b. Use the fake CA to sign the admin CSR, yielding a certificate.
+    openssl x509 -req -CA cacert.pem -CAkey cakey.pem \
+    -copy_extensions copy -days "${DAYS}" \
+    -in admin-req.pem -out admin-cert.pem 2> /dev/null
+    echo "Created admin certificate"
+
+    rm -f node-req.pem admin-req.pem
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:476]: These are not certificates.  These are CSRs, certificate signing requests.  They're useless once the certificates are issued, which happens above in 2b and 3b.

File: opensearch/sycamore-opensearch.sh [hunk 16]
```diff
@@ -427,4 +433,61 @@ setup_transient() {
     deploy_model "${OPENAI_MODEL_ID}" "${OPENAI_TASK_ID}" "OpenAI"
 }
 
+create_certificates() {
+    local HOST="${SSL_HOSTNAME:-localhost}"
+    local DAYS=10000
+
+    pushd config
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:440]: Can do.

File: opensearch/sycamore-opensearch.sh [hunk 17]
```diff
@@ -83,19 +77,31 @@ die() {
...
+    local out=$(_curl http://localhost:9200/)
+    [[ -z ${out} ]] && return 1
+    local name=$(jq -r '.name' <<< ${out})
+    [[ -z ${name} ]] && return 1
+    return 0
+}
+
+opensearch_up_ssl() {
+    local out=$(_curl https://localhost:9200/)
+    [[ -z ${out} ]] && return 1
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:90]: Yes, double square brackets cause the parsing to be more semantic; there's no word splitting.

File: opensearch/sycamore-opensearch.sh (hunk 2 above)
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:490]: This survives restarts.  It would be pretty lousy security otherwise.

While the aryn_ssl_setup_complete semaphore isn't used in this script, it is used by stuff in pyscripts.  It's included here for consistency and semantic completeness.  After this milestone is when it's safe to register/deploy models.  If this is absent, it's a useful debugging clue.

File: opensearch/roles_mapping.yml (hunk 3 above)
Comment (alexaryn) [reply opensearch/roles_mapping.yml:7]: Both 'all_access' and 'opendistro_security_anonymous_backendrole' seem to be supplied out of the box.  Here they're linked together to achieve the same behavior we had without the security plugin.  The current task is to communicate over SSL.  We can create another task for passwords once we figure out how we want to do user management and authentication.  OpenSearch seems to prefer using client certificates over username/password.

File: opensearch/Dockerfile (hunk 4 above)
Comment (alexaryn) [reply opensearch/Dockerfile:29]: It wasn't explicit in that version, as it was inherited from the parent image.  That seemed error-prone.  Things are rearranged now and I'll try without it.

File: opensearch/Dockerfile (hunk 5 above)
Comment (alexaryn) [reply opensearch/Dockerfile:32]: OpenSearch complains when these files' modes are too permissive.

File: opensearch/sycamore-opensearch.sh [hunk 18]
```diff
@@ -83,19 +77,31 @@ die() {
...
+    local name=$(jq -r '.name' <<< ${out})
+    [[ -z ${name} ]] && return 1
+    return 0
+}
+
+opensearch_up_net() {
+    _curl http://localhost:9200/ -o /dev/null
+    [[ ($? != 0) && ($? != 52) ]] && return 1
```
[the remark this answers is not in the document]
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:98]: I'll add a comment.  You get 52 if OpenSearch wants SSL.

File: opensearch/sycamore-opensearch.sh (hunk 1 above)
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:483]: I'll add a link to the doc I took this from.

**Review by eric-anderson — COMMENTED** [review]
Going to need to go through docs and search for "use_ssl": False after this change and fix it.  Example will be running_a_data_preparation_job.md after https://github.com/aryn-ai/sycamore/pull/233

File: opensearch/sycamore-opensearch.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (eric-anderson) [reply opensearch/sycamore-opensearch.sh]: Yes, but it seems to me like a good idea to check to make sure someone doesn't incorrectly change it to persistent.  If it's ever there, something has gone very very wrong (and the check is cheap).

File: jupyter/Dockerfile [hunk 19]
```diff
@@ -5,12 +5,13 @@ FROM arynai/sycamore-importer:$TAG
...
 # TODO: eric - remove this once we switch other containers over to non-root
 # Note user/group should be 1000 to match with likely defaults on hosts.
+USER root
 COPY jupyter/docker-root-setup.sh /root
 RUN /bin/bash /root/docker-root-setup.sh
 COPY jupyter/sudoers /etc/sudoers
 COPY jupyter/fixuser.py /root/fixuser.py
 
-USER 1000
+USER app
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply jupyter/Dockerfile:15]: Unfixed?

File: importer/docker_local_import.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (eric-anderson) [reply importer/docker_local_import.py:377]: `# Leave verification off since we are using self-signed certificates`

File: jupyter/run-jupyter.sh (hunk 10 above)
[the remark this answers is not in the document]
Comment (eric-anderson) [reply jupyter/run-jupyter.sh]: `# Use RSA as the more commonly used option`

File: jupyter/run-jupyter.sh [hunk 20]
```diff
@@ -16,6 +16,17 @@ ln -snf "${JUPYTER_CONFIG_DOCKER}" $HOME/.jupyter
...
 rm /app/.local/share/jupyter/runtime/jpserver-*-open.html 2>/dev/null
 
+: ${HOST:=localhost}
+SSLPFX="/app/work/docker_volume/${HOST}"
+if [[ (! -f ${SSLPFX}-key.pem) || (! -f ${SSLPFX}-cert.pem) ]]; then
+    openssl req -batch -x509 -newkey rsa:4096 -days 10000 \
+    -subj "/C=US/ST=California/O=Aryn.ai/CN=${HOST}" \
+    -extensions v3_req -addext "subjectAltName=DNS:${HOST}" \
+    -noenc -keyout "${SSLPFX}-key.pem" -out "${SSLPFX}-cert.pem" \
+    2> /dev/null
```
Comment (eric-anderson) [inline jupyter/run-jupyter.sh:26]: We should check for an error/save the logs in the openssl call. The same || die construction from sycamore-opensearch.sh (and copy over die)

File: examples/simple_config.py [hunk 21]
```diff
@@ -27,7 +27,7 @@
     "hosts": [{"host": "localhost", "port": 9200}],
     "http_compress": True,
     "http_auth": ("admin", "admin"),
-    "use_ssl": False,
+    "use_ssl": True,
```
Comment (eric-anderson) [inline examples/simple_config.py:30]: Looks like there are a few more of these hidden in the documentation.
`grep -R use_ssl.*False .` finds 4

File: jupyter/Dockerfile [hunk 22]
```diff
@@ -5,19 +5,21 @@ FROM arynai/sycamore-importer:$TAG
...
+# User 'app' must be id number 1000 (the default) for all to work properly.
+USER app
 RUN poetry run pip install notebook ipywidgets 'opensearch-py>=2.4' && rm -rf /app/.cache
 RUN mkdir -p /app/work/docker_volume /app/work/bind_dir /app/work/examples
 RUN touch /app/work/AAA_SEE_README_FOR_PERSISTENT_DATA_DIRECTORIES
 COPY jupyter/run-jupyter.sh ./
 COPY jupyter/README.md ./work
-COPY sycamore_local_dev_example.ipynb ./work/examples
-RUN perl -i -pe 's/localhost/opensearch/ if /9200/;s,tmp/sycamore/data,/app/work/docker_volume,' /app/work/examples/sycamore_local_dev_example.ipynb
+COPY notebooks/jupyter_dev_example.ipynb ./work/examples
```
Comment (eric-anderson) [inline jupyter/Dockerfile:21]: This would have been a good thing to have in another PR; looks like we broke Docker builds at head accidentally during the merge.

File: opensearch/Dockerfile (hunk 5 above)
Comment (eric-anderson) [reply opensearch/Dockerfile:36]: `# Permissions must be restricted for opensearch to be happy`
It's not the number of layers it's the size of the layers.  You might think that when you just chown a file all that would be recorded would be a metadata change, but (to my surprise) Docker records the entire file. As a result our Jupyter image is about 2x the size it needs to be since the sycamore image has the files as root and the Jupyter one has them as 1000.  Given it does that for chown, I assume it does the same thing for chmod.  As I said, it doesn't really matter given the rest of our size so just adding the comment is sufficient, but it's worth other people understanding the Docker gotchas.

File: opensearch/opensearch.yml [hunk 23]
```diff
@@ -3,3 +3,13 @@ http.cors.enabled : true
 http.cors.allow-origin: "*"
 http.cors.allow-methods: OPTIONS,HEAD,GET,POST,PUT,DELETE
 http.cors.allow-headers: X-Requested-With,X-Auth-Token,Content-Type,Content-Length
+plugins.security.ssl.transport.enforce_hostname_verification: false
```
Comment (eric-anderson) [inline opensearch/opensearch.yml:6]: If you can comment on the false here as well that would be nice. I don't know if yml has comments.
`# Self-signed certificates so no hostname verification`

File: opensearch/sycamore-opensearch.sh (hunk 2 above)
Comment (eric-anderson) [reply opensearch/sycamore-opensearch.sh:502]: Why not use aryn_deploy_complete in pyscripts?
`# semaphore for debugging state if something goes wrong; can be used as a signal that it is ok to deploy models`

**Review by eric-anderson — CHANGES_REQUESTED** [review]

File: opensearch/sycamore-opensearch.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh]: My feeling is that we're getting out on the long tail of probability here.  A comment saying "It's important to keep this transient to match the lifetime of model deployment" should be sufficient.  If we need to add a check for this, there are so many other aspects of this script that we'd want to check to make sure someone didn't incorrectly change them (e.g.: pipeline names).  The script is already hard to follow and I don't think loading it up with checks will help that.

File: jupyter/Dockerfile (hunk 19 above)
[the remark this answers is not in the document]
Comment (alexaryn) [reply jupyter/Dockerfile:15]: Fixed.  This diff view seems frozen in time.

File: opensearch/sycamore-opensearch.sh (hunk 2 above)
Comment (alexaryn) [reply opensearch/sycamore-opensearch.sh:502]: The place in pyscripts that uses this is waiting for a different milestone in startup.  It's actually waiting for OpenSearch to be ready to register and deploy models.  I'll tweak the comment.

File: opensearch/sycamore-opensearch.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (eric-anderson) [reply opensearch/sycamore-opensearch.sh]: I figured ```
[[ $(_curl "${BASE_URL}/_cluster/settings" | grep -c aryn_deploy_complete) == 0 ]] || die "aryn_deploy_complete already there"

Was pretty simple and it prevents against opensearch not behaving the way we expect (I've had numerous issues with things like this in the past). 

Would rewriting the script in python make it easier to follow (not in this PR)?  My threshold for bash script rewriting is usually "when it gets too complicated", and maybe we've reached that point.

File: jupyter/Dockerfile (hunk 19 above)
[the remark this answers is not in the document]
Comment (eric-anderson) [reply jupyter/Dockerfile:15]: Did you not push? at commit f0f484a47813057cc92c03ada6f2612848dadec4 I see:
```
# User 'app' must be id number 1000 (the default) for all to work properly.
USER app
```
I was expecting this to be:
```
# App user must be id number 1000 (the default) for all to work properly.
USER 1000
```

To make sure that files are installed as the correct user id.

**Review by eric-anderson — APPROVED** [review]
